### ISSUE TITLE
Datepicker component uses non-accessible Flatpickr calendar dropdown

### ISSUE THREAD

**Issue by dakahn** [IBM/carbon-components-react#1899]
### Problem

The Datepicker relies on the inaccessible Flatpickr component to display a calendar dropdown. This calendar dropdown has [strange or missing keyboard navigation](https://github.com/flatpickr/flatpickr/issues/1636) and is being [flagged by our automated accessibility testing system](https://github.com/IBM/carbon-components/issues/1471) (DAP). 

### Solution

Two possible paths forward: we can submit a PR to Flatpickr, but given the release schedule and that work on their end [has slowed](https://github.com/flatpickr/flatpickr/issues/1646) a better long term (though more painful potentially short term) solution may be to convert our date input to a[ native html 5 calendar dropdown](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date).

**Comment by bakerTX** [comment]
We're also having a _lot_ of trouble with this component. Not only is the datepicker throwing DAP violations, but it straight up **trapping** keyboard focus: When using the 'range with calendar' option, it seems impossible to tab out of the input in the forward direction.

trying to tab forward:
![feb-25-2019 16-34-27](https://user-images.githubusercontent.com/8573243/53373508-4bcae880-391b-11e9-828e-e241a311db53.gif)

**Comment by chmln** [comment]
Hey guys, flatpickr author here.

@bakerTX I can't fix problems that are not reported. This is especially pertinent to plugins that have a "beta" status. I have gone ahead and fixed the trapping issue. The patched version is available right now (release v4.5.5).

@dakahn this sounds disingenuous. The accessibility issue you've opened not only contains incorrect information but is not actionable at all. The project has always been receptive to pull requests; and contrary to your tone is very well maintained with 4 developers onboard at this point.

**Comment by asudoh** [comment]
Hi @chmln :wave: thank you so much for chiming in! And our apologies that the issue description here may have been too general against Flatpickr. In fact, though I wasn't on Carbon team when Carbon started using Flatpickr, I could see at then that Flatpickr choice for our date picker allows us to develop the component very quickly meeting our design requirement as well as g11n requirements at that time.

OTOH our design/UX requirements around range mode date picker and its difference from what Flatpickr provides out-of-the-box (or with a plugin) have required some tweaks wrt how our code interacts with Flatpickr. Most of such portion has been gone by using range plugin, though as you mentioned it's in "beta" status. But there are still some, and the combination of those yielded several issues opened still, where:

* It's hard to create reduced test cases to report issues to Flatpickr repo, as many issues happen only with the integration of Flatpickr with our date picker code and it takes non-trivial efforts wrt what portion/requirements in our code caused such integration issues
* Those bring up a lingering question if Flatpickr fits (fully) well in our specific set of use cases

BTW, I felt very grateful when I could see you quickly fixed what @bakerTX mentioned in above comment! As your referral to non-reported issues seems to allude to, better communnication on specifics what issues we have would be worthwhile going forward.

**Comment by chmln** [comment]
@asudoh thanks for clearing things up. 

> It's hard to create reduced test cases to report issues to Flatpickr repo, as many issues happen only with the integration of Flatpickr with our date picker code and it takes non-trivial efforts wrt what portion/requirements in our code caused such integration issues

If you guys ever run into specific problems like @bakerTX did, please file those immediately - bugs are usually fixed rather quickly. 

Also, it sounds like the project could benefit a lot from having certain changes upstreamed. The best way to do that is to file issues for bugs and missing features. Doing so will not only ease the maintenance burden, but also make things easier for other users and projects like Carbon. 

> BTW, I felt very grateful when I could see you quickly fixed what @bakerTX mentioned in above comment! As your referral to non-reported issues seems to allude to, better communnication on specifics what issues we have would be worthwhile going forward.

You're most welcome :+1:

**Comment by bakerTX** [comment]
Hey @chmln and Carbon folks. I think I could have been more clear. I don't think the issue starts with the Flatpickr component. I was not able to reproduce the focus trapping issue on the Flatpickr examples page, only on the Carbon React storybook and in our implementation of that. In Carbon's datepicker.js, there is a lot of extra logic added onto the Flatpickr dependency. I'm suspicious that the issue will actually be found there, and not in Flatpickr source code. Like @asudoh said, I'm very grateful for the quick fix! But I suspect we will still require some patching-up in the Carbon source code to get this component to a working state.

**Comment by dakahn** [comment]
@chmln Firstly, I apologize if my issue came off as a provocation. Enough time had passed after opening that from my standpoint it seemed like either nobody had seen it or it wasn't a priority. Bummed to hear it was neither and you don't think there's anything actionable. I've added a response for clarity on user expectations and keyboard interaction conventions when it comes to date pickers. Happy to continue that discussion.

And hey, I can totally be wrong -- but I try not to be wrong on purpose 😺. It looks like there are undocumented keyboard affordances for controlling month to month switching -- that's great and thanks for your response. I've posted in the original issue about how these controls are maybe problematic and unintuitive. If it's within the scope of your project maybe we can collaborate to make these better. ☮️

**Comment by dakahn** [comment]
I'm closing this as we've decided to continue forward with Flatpickr and submit PRs to their repo getting us where we need to be.
